README.md
# wiring_analog_SAMD_TT

## Library with changes to SAMD wiring_analog.c and ADC calibration code added

This library has two pairs of files:

**wiring_analog_SAMD_TT.h and .c:** These provide functions that directly replace those in the Arduino SAMD core file named *wiring_analog.c*, to fix some problems and provide some new features I needed.

**calibSAMD_ADC_withPWM.h and .c:** These provide a function to calibrate the ADC using least squares regression on multiple samples of a reliable set of voltages created using a simple resistor-capacitor external circuit together with a pulse-width-modulation signal created by an internal microprocessor timer.

## Changes to wiring_analog.c in wiring_analog_SAMD_TT.c

The changes made to *wiring_analog.c* within *wiring_analog_SAMD_TT.c* are:

**1.** Header file <wiring_analog_SAMD_TT.h> is provided to declare the .c functions.

**2.** Every function in *wiring_analog_SAMD_TT.c* has the same name as the original in *wiring_analog.c* but with
  "_SAMD_TT" appended, e.g. *analogRead()* becomes *analogRead_SAMD_TT()*. This allows both files to be used at the same time with minimal interference, if necessary.
  
**3.** The original *analogWrite()* function, when writing to a digital port, mapped the analog output value to 8-bits and then wrote LOW if < 128 else HIGH. The new analogWrite_SAMD_TT() writes LOW if the value is 0 and otherwise HIGH, following the usual C and C++ interpretation of an integer as being false only when it is 0.

**4.** The original *analogRead()* function didn't accept use of D4/D5/D6/D7 as analog inputs, even though they can be used that way on the SAMD21G18 microprocessor. Instead, it mapped those to A4/A5/A6/A7. The new function *analogRead_SAMD_TT()* removes that mapping and allows those pins to be used as analog inputs.

**5.** The original *analogRead()* function would disable the DAC when it had been previously written (and thus enabled) for the same pin as is being read, which may not be what is wanted if the DAC is being used to drive the ADC pin for calibration and you desire to read with the ADC the voltage that the DAC generates. New function *analogRead_disableDACoption_SAMD_TT()* is a copy of analogRead() with one additional bool argument named 'disableDAC'. The argument value is used to decide whether or not to disable the DAC. The new *analogRead_SAMD_TT()* function simply calls that function with the 'disableDAC' argument set to true.

**6.** The write resolution for PWM operations was shared with the write resolution for the DAC, and either could be changed at any time. This code separates the write resolutions, adding the functions *analogWriteResolution_DAC_SAMD_TT()* to set DAC resolution and *analogWriteResolution_PWM_SAMD_TT()* to set PWM resolution. The new *analogWriteResolution_SAMD_TT()* function sets BOTH of the new resolutions to maintain compatibility with *analogWriteResolution()*.

**7.** The original *analogWrite()* function had some problems with regard to using a timer for pulse width modulation:

> Some TCC timers are 16-bit and some are 24-bit and there is no function to tell you how many bits a given pin's timer has.
>
> It used a fixed value of 0xFFFF for the waveform period with a TCC timer even though some TCC timers are 16-bit and some are 24-bit.
>
> Often the user will want to vary the PWM period but this was not possible.
>
> The write resolution was not applied to the period value.
>
> *analogWrite()* both initialized the timer for PWM operations the first time it was called for each timer AND loaded it with the (fixed) pulse period and on-time values. There was no way to reinitialize the timer.

To solve these issues, these functions were added:

> **analogGetResolution_TCC_SAMD_TT():** return the resolution of a TCC.
>
> **analogStartPWM_TCC_SAMD_TT():** initialize a timer for PWM.
>
> **analogSetPWM_TCC_SAMD_TT():** write on-time and period values to the PWM, honoring PWM resolution when needed.
  
The PWM functionality in the **analogWrite_SAMD_TT()** function remains unchanged from *analogWrite()*.

### Example in wiring_analog_SAMD_TT.ino

See the file *wiring_analog_SAMD_TT.ino* in the library folder's examples subfolder for a program that calls several functions in *wiring_analog_SAMD_TT.c* to illustrate usage. The program also calls *calibSAMD_ADC_withPWM()* in *calibSAMD_ADC_withPWM.cpp* to calibrate the ADC, which means an external circuit with two capacitors and a resistor is required (see next section of this README file for a description). The calibration uses three port pins on the microcomputer, and the subsequent calls to *wiring_analog_SAMD_TT.c* functions use those same pins and one more, in order to illustrate writing to the digital-to-analog converter (DAC), configuring a TCC timer for pulse width modulation (PWM) and writing different PWM periods and duty cycles, and reading from the analog-to-digital converter (ADC). Once a second, a line is printed to the Arduino IDE serial monitor window show the voltage values read from PIN_ADC by the ADC along with the expected voltages produced on PIN_ADC by the PWM. An oscilloscope, if available, will show the voltages and PWM signals that are generated on three pins (PIN_DAC, PIN_PWM, and PIN_ADC) by this program.

## New ADC calibration function in calibSAMD_ADC_withPWM.h and .c

I've observed poor performance from the SAMD21G18A Nano 33 IoT analog-to-digital converter (ADC), even after using the calibration procedure in open-source program CorrectADCResponse.ino. That program assumes that 0 and 3.3V are the two voltage-ends of the ADC, but the ADC seems to have problems at the ends of its reference voltage range, at least in an earlier Nano I had. Subsequent Nanos have been better, though still not great. There is a lot of noise visible in the values read from the ADC. Turning on the averaging mode to average around 64 samples automatically helps a lot. However, I found that I could further improve the performance by recomputing the ADC gain and offset calibration parameters using least squares regression, which is the purpose of function calibSAMD_ADC_withPWM() in files calibSAMD_ADC_withPWM.h/.cpp.

The new calibration function is named **calibSAMD_ADC_withPWM()**. It computes the gain and offset parameters using the ADC and a TCC timer with pulse width modulation (PWM) and an external circuit consisting of the PWM output fed into a resistor and capacitor in series, used to generate reliable voltages between GND and 3.3V that are fed to the ADC analog input. A series of stepped voltages are converted using the ADC, and least squares regression is applied to the actual and expected results to obtain gain and offset correction values. These are loaded into the ADC registers before the function returns.

*12-bit ADC resolution and external voltage reference is set by the function and required for all ADC operation.*

This code is customized for use in a SAMD microcomputer, such as in the Arduino Nano 33 IoT. It won't work for non-SAMD systems.

### Required hardware and ports

This requires use of the following hardware and ports:

**1.** The internal analog-to-digital converter device and one ADC analog input (pinADC)

**2.** A TCC timer device and one digital output (pinPWM) that can be driven by the timer

**3.** A digital output (pin_AREF_OUT) that is tied to the AREF input for generating the ADC analog reference voltage in a manner that can be turned on and off. This is optional -- you can do away with this requirement.

**4.** Optional digital output (PIN_DEBUG) used for debugging, disableable.

### Required circuit wiring

You must provide the following circuitry on your microcontroller system:

**1.** Wire a 0.1 uF capacitor between GND and pinADC.

**2.** Wire a 10K ohm resistor between pinADC and pinPWM.

**3.** Wire a large (say 100 uF) capacitor between GND and pinAREF_OUT.

**4.** Wire pinAREF_OUT to the AREF (external analog reference voltage) pin.

**5.** Optionally, during testing you can connect PIN_DEBUG to an oscilloscope input to trigger the scope when events happen that affect pinADC and/or pinPWM. See CALIB_ADC_SCOPE.

Alternatively, wire the AREF pin directly to +3.3V and don't use pinAREF_OUT. Then delete the code herein that configures and drives pinAREF_OUT.

The pins used are defined by pin* arguments to calibSAMD_ADC_withPWM(). These are the standard "Arduino pin numbers", which are not really pin numbers at all. A "pin" number is actually an index into table g_APinDescription[], which can be found in SAMD core software file variant.cpp.

### To incorporate ADC calibration into your program

I incorporated this hardware into my thermostat system, and the calibration code runs each time the thermostat starts up. To incorporate ADC calibration into your program:

**1.** Choose pin constants for the pins you wish to use for pinADC, pinPWM, and pinAREF_OUT (optional).

**2.** Choose a value for cfgADCmultSampAvg, which determines whether or not automatic ADC multi-sampling and averaging is done. It works well and is recommended.

**3.** #include <calibSAMD_ADC_withPWM.h> in your program.

**4.** Call calibSAMD_ADC_withPWM() from setup() before using the ADC, passing it the appropriate 'pin' values and 'cfgADCmultSampAvg'.

**5.** Wire your system as shown above.

**6.** Compile your program and download it into a SAMD-based microcomputer.

**7.** When calibSAMD_ADC_withPWM() runs, it will compute and load gain and offset corrections into the ADC. Subsequent ADC reads should be more accurate.

*This requires use of the file named wiring_analog_SAMD_TT.c, which is part of this library.*

### To use debugging features to check out ADC calibration hardware and operation

*calibSAMD_ADC_withPWM()* optionally uses the library module named monitor.printf(). If you enable it by setting CALIB_ADC_DBG to 1 and calling monitor.begin() from your setup() function, calibSAMD_ADC_withPWM() will print out debugging information on the serial monitor of the Arduino IDE. Look at the values printed for gainError and offsetError and ADC values that you can check to make sure the function is working properly.

You can copy and paste ADC output values into the spreadsheet file "calibSAMD_ADC_withPWM.xlsx" (extras subfolder within the library folder) into one of the data columns to plot the input/output error curve.

I used an oscilloscope to debug this module. If you have a scope, you can enable the debugging code I used and monitor pinPWM and pinADC and PIN_DEBUG with it to check out the external PWM circuit. See CALIB_ADC_SCOPE.

You may find through debugging that you need to adjust certain constants:

> AREF_STABLE_DELAY
>
> PWM_STABLE_DELAY
>
> PERCENT_AT_ENDS

Most likely, the values already set will work fine, if you use the recommended capacitor and resistor values.

### Example in calibSAMD_ADC_withPWM.ino

See the file *calibSAMD_ADC_withPWM.ino* in the library folder's examples subfolder for a program that calls calibSAMD_ADC_withPWM() during its setup() function.

## Finding variant.cpp and other system files

Here are some ways to view and get information about system files.

Right-click on "#include <Arduino.h>" in Arduino IDE and choose "Go to Definition" to open the file in the IDE. From there, open WVariant.h, variant.h, and samd.h. In WVariant.h, right-click on "extern const PinDescription g_APinDescription[];" and choose "Go to Definition" to open variant.cpp. From samd.h, open samd21.h, and from there, open samd21g18a.h, and from there open both the instance and component .h files for the peripherals of interest, say adc.h and tcc.h. To open wiring_analog.c, insert a call to analogWrite(), right-click and choose "Go to Definition". In that file, find a call to pinPeripheral and right-click it and choose "Go to Definition" to open wiring_private.c, and look at that function to see what it does. In samd21g18a.h note the definitions of symbols like ADC and DAC (the second definition, not the first). To find the path of an open file, hover over its name in Arduino IDE open tab for that file. Also open wiring_analog_SAMD_TT.cpp and compare to wiring_analog.c.

## SAMD21G18 microprocessor datasheet

The Nana 33 IoT microcomputer uses an Atmel SAMD21G18 microprocessor. You can download the data sheet by going to:

> https://www.microchip.com/en-us/product/ATsamd21g18

and downloading the data sheet PDF (over 1000 pages, daunting!) To use it, study table 7-1 and note how g_APinDescription[] in variant.cpp uses its info. Then, for each peripheral of interest, in the Table of Contents at the top, find the chapter for that peripheral. You can read the details of the peripheral, or go to the "Register Summary" for it, click on registers and read their description. The .h files mentioned above define all registers of all peripherals in a clean easy-to-use manner. E.g., ADC CTRLB register can be accessed with ADC->CTRLB.reg, and the PRESCALER field in that register can be accessed with ADC->CTRLB.bit.PRESCALER.

## Contact

If you find a problem, please submit an issue report [here](https://github.com/tedtoal/wiring_analog_SAMD_TT/issues/new/choose).

You can contact me (the author) at [ted@tedtoal.net](ted@tedtoal.net)


CHANGELOG.md
# wiring_analog_SAMD_TT change log
---

## Table of Contents

* [Release 1.1.1](#release-111)
---

### Release 1.1.1

1. Initial release.


src/wiring_analog_SAMD_TT.h
/*
  wiring_analog_SAMD_TT.h - Provides the same functions as SAMD wiring_analog.c
  with several changes to fix problems.

  Created by Ted Toal, July 21, 2023, declaring the functions that are defined
  in wiring_analog_SAMD_TT.c.

  Copyright (c) 2014 Arduino LLC.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

  Brief summary of changes to wiring_analog_SAMD_TT.c from wiring_analog.c:
    1. File <wiring_analog_SAMD_TT.h> is provided to declare the .c functions.
    2. Every function in this file has the same name as the original but with
      "_SAMD_TT" appended, e.g. analogRead() becomes analogRead_SAMD_TT().
    3. The original analogWrite(), when writing to a digital port, mapped the
      analog output value to 8-bits and then wrote LOW if < 128 else HIGH. The
      new analogWrite_SAMD_TT() writes LOW if the value is 0 and otherwise HIGH.
    4. The original analogRead() function doesn't accept use of D4/D5/D6/D7
      as analog inputs, even though they can be used that way. Instead, it maps
      those to A4/A5/A6/A7. analogRead_SAMD_TT() removes that mapping and allows
      those pins to be used as analog inputs.
    5. The original analogRead() function would disable the DAC when it has been
      previously written (and thus enabled) for the same pin as is being read,
      which may not be what is wanted if the DAC is being used to drive the ADC
      pin for calibration. Function analogRead_disableDACoption_SAMD_TT() is
      added, that is a copy of analogRead() with one additional bool argument
      named 'disableDAC'. The argument value is used to decide whether or not to
      disable the DAC. The analogRead_SAMD_TT() function simply calls that
      function with the 'disableDAC' argument set to true.
    6. The write resolution for the PWM was shared with that for the DAC, and
      could be changed at any time. This code separates the write resolutions,
      adding the functions analogWriteResolution_DAC_SAMD_TT() and
      analogWriteResolution_PWM_SAMD_TT(). The analogWriteResolution_SAMD_TT()
      function sets BOTH of the new resolutions to maintain compatibility with
      analogWriteResolution().
    7. The original analogWrite() function had some problems with regard to
      using a timer for pulse width modulation:
        - Some TCC timers are 16-bit and some are 24-bit and there is no
          function to tell you how many bits a given pin's timer has.

        - It used a fixed value of 0xFFFF for the waveform period with a TCC
          timer even though some TCC timers are 16-bit and some are 24-bit.

        - Often the user will want to vary the PWM period but this was not
          possible.

        -  The write resolution was not applied to the period value.

        - analogWrite() both initialized the timer for PWM operations the first
          time it was called for each timer AND loaded it with the (fixed) pulse
          period and on-time values. There was no way to reinitialize the timer.

      To solve these issues, these functions were added:
        analogGetResolution_TCC_SAMD_TT(): return the resolution of a TCC.
        analogStartPWM_TCC_SAMD_TT(): initialize a timer for PWM.
        analogSetPWM_TCC_SAMD_TT(): write on-time and period values to the PWM,
          mapping the values for PWM resolution when necessary.

      These use uint32_t type for value and period rather than int type.
      The PWM functionality in the analogWrite_SAMD_TT() function remains
      unchanged.
*/
#ifndef wiring_analog_SAMD_TT_h
#define wiring_analog_SAMD_TT_h

#include <Arduino.h>

#ifdef __cplusplus
extern "C" {
#endif

extern void analogReadResolution_SAMD_TT(int res);
extern void analogWriteResolution_SAMD_TT(int res);
extern void analogWriteResolution_DAC_SAMD_TT(int res);
extern void analogWriteResolution_PWM_SAMD_TT(int res);
extern void analogReference_SAMD_TT(eAnalogReference mode);
extern int analogRead_SAMD_TT(pin_size_t pin);
extern int analogRead_disableDACoption_SAMD_TT(pin_size_t pin, bool disableDAC);
extern void analogWrite_SAMD_TT(pin_size_t pin, int value);
extern int analogGetResolution_TCC_SAMD_TT(pin_size_t pin);
extern int analogStartPWM_TCC_SAMD_TT(pin_size_t pin);
extern bool analogSetPWM_TCC_SAMD_TT(pin_size_t pin, int res, uint32_t on_time,
  uint32_t period);

#ifdef __cplusplus
}
#endif

#endif // wiring_analog_SAMD_TT_h


src/calibSAMD_ADC_withPWM.h
/*
  calibSAMD_ADC_withPWM.h - a function to calibrate the SAMD ADC.
  Created by Ted Toal, May 21, 2023.
  Released into the public domain.

  Software License Agreement (BSD License)

  Copyright (c) 2023 Ted Toal
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:
  1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.
  3. Neither the name of the copyright holders nor the
  names of its contributors may be used to endorse or promote products
  derived from this software without specific prior written permission.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ''AS IS'' AND ANY
  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


  ******************************************************************************
  Details:

  I've observed poor performance from the SAMD21G18A Nano 33 IoT ADC, even after
  using the calibration procedure in open-source program CorrectADCResponse.ino.
  That program assumes that 0 and 3.3V are the two voltage-ends of the ADC, but
  the ADC seems to have problems at the ends of its reference voltage range, at
  least in an earlier Nano I had. Subsequent Nanos have been better, though
  still not great, with a lot of noise visible in the values read from the ADC.
  Turning on the averaging mode to average around 64 samples automatically helps
  a lot. I found that I could further improve the performance by recomputing the
  ADC gain and offset calibration parameters.

  This module defines function calibSAMD_ADC_withPWM() that computes those
  parameters using the ADC and a TCC timer with pulse width modulation (PWM) and
  an external circuit consisting of the PWM output fed into a resistor and
  capacitor in series, used to generate reliable voltages between GND and 3.3V
  that are fed to the ADC analog input. The new gain and offset values are
  loaded into the ADC registers before the function returns.

  NOTE: 12-bit ADC resolution and external voltage reference is set and required.

  This code is customized for use in a SAMD microcomputer, such as in the
  Arduino Nano 33 IoT. It won't work for non-SAMD systems.

  This requires use of the following hardware and ports:
    1. The ADC and one ADC analog input (pinADC)
    2. A TCC timer and one digital output (pinPWM) that can be driven by the timer
    3. A digital output (pinAREF_OUT) that is tied to the AREF input for
      generating the ADC analog reference voltage in a manner that can be turned
      on and off.
    4. Optional digital output (PIN_DEBUG) used for debugging, disableable.

  Circuit wiring:
   1. Wire a 0.1 uF capacitor between GND and pinADC.
   2. Wire a 10K ohm resistor between pinADC and pinPWM.
   3. Wire a large (say 100 uF) capacitor between GND and pinAREF_OUT.
   4. Wire pinAREF_OUT to the AREF (external analog reference voltage) pin.
   5. Optionally, during testing you can connect PIN_DEBUG to an oscilloscope
      input to trigger the scope when events happen that affect pinADC and/or
      pinPWM. See CALIB_ADC_SCOPE below.

  Alternatively, wire the AREF pin directly to +3.3V and don't use pinAREF_OUT.
  (Delete the code herein that configures and drives pinAREF_OUT).

  The pins used are defined by pin* arguments to calibSAMD_ADC_withPWM(). These
  are the standard "Arduino pin numbers", which are not really pin numbers at
  all. A "pin" number is actually an index into table g_APinDescription[], which
  can be found in SAMD core software file variant.cpp.

  NOTES ON FINDING variant.cpp and other system files:
  Right-click on "#include <Arduino.h>" in Arduino IDE and choose "Go to
  Definition" to open the file in the IDE. From there, open WVariant.h,
  variant.h, and samd.h. In WVariant.h, right-click on "extern const
  PinDescription g_APinDescription[];" and choose "Go to Definition" to open
  variant.cpp. From samd.h, open samd21.h, and from there, open samd21g18a.h,
  and from there open both the instance and component .h files for the
  peripherals of interest, say adc.h and tcc.h. To open wiring_analog.c,
  insert a call to analogWrite(), right-click and choose "Go to Definition".
  In that file, find a call to pinPeripheral and right-click it and choose "Go
  to Definition" to open wiring_private.c, and look at that function to see
  what it does. In samd21g18a.h note the definitions of symbols like ADC and
  DAC (the second definition, not the first). To find the path of an open
  file, hover over its name in Arduino IDE open tab for that file. Also open
  wiring_analog_SAMD_TT.cpp and compare to wiring_analog.c.

  NOTE ON MICROPROCESSOR DATASHEET: download the data sheet by going to:
  https://www.microchip.com/en-us/product/ATsamd21g18
  and downloading the data sheet PDF (over 1000 pages, daunting!) To use it,
  study table 7-1 and note how g_APinDescription[] in variant.cpp uses its
  info. Then, for each peripheral of interest, in the Table of Contents at the
  top, find the chapter for that peripheral. You can read the details of the
  peripheral, or go to the "Register Summary" for it, click on registers and
  read their description. The .h files mentioned above define all registers of
  all peripherals in a clean easy-to-use manner. E.g., ADC CTRLB register can
  be accessed with ADC->CTRLB.reg, and the PRESCALER field in that register
  can be accessed with ADC->CTRLB.bit.PRESCALER.

  I incorporated this hardware into my thermostat system, and the calibration
  code runs each time the thermostat starts up. To incorporate ADC calibration
  into your program:

    1. Choose pin constants for the pins you wish to use for pinADC, pinPWM,
        and pinAREF_OUT (optional).
    2. Choose a value for cfgADCmultSampAvg, which determines whether or not
        automatic ADC multi-sampling and averaging is done. It works well and
        is recommended.
    3. #include <calibSAMD_ADC_withPWM.h> in your program.
    4. Call calibSAMD_ADC_withPWM() from setup() before using the ADC, passing
        it the appropriate 'pin' values and 'cfgADCmultSampAvg'.
    5. Wire your system as shown above.
    6. Compile your program and download it into a SAMD-based microcomputer.
    7. When calibSAMD_ADC_withPWM() runs, it will compute and load gain and
        offset corrections into the ADC. Subsequent ADC reads should be more
        accurate.

  This requires use of the file named wiring_analog_SAMD_TT.c, which is part of
  the library containing this file.

  This optionally uses the library module named monitor.printf(). If you enable
  it below by setting CALIB_ADC_DBG to 1 and calling monitor.begin() from your
  setup() function, calibSAMD_ADC_withPWM() will print out debugging information
  on the serial monitor of the Arduino IDE. Look at the values printed for
  gainError and offsetError and ADC values that you can check to make sure the
  function is working properly. You can copy and paste ADC output values into
  the spreadsheet file "calibSAMD_ADC_withPWM.xlsx" into one of the data columns
  to plot the input/output curve.

  I used an oscilloscope to debug this module. If you have a scope, you can
  enable the debugging code I used and monitor pinPWM and pinADC and PIN_DEBUG
  with it to check out the external PWM circuit. See CALIB_ADC_SCOPE below.

  You may find through debugging that you need to adjust certain constants below:
    AREF_STABLE_DELAY
    PWM_STABLE_DELAY
    PERCENT_AT_ENDS
  ******************************************************************************
*/

#ifndef calibSAMD_ADC_withPWM_h
#define calibSAMD_ADC_withPWM_h

#include <Arduino.h>

// Set this to 1 to #include monitor_printf.h and print debug results of the
// calibration procedure to the Arduino IDE monitor when calibSAMD_ADC_withPWM()
// is called. If set to 1, you must call monitor.begin() from setup() BEFORE
// calling calibSAMD_ADC_withPWM().
#define CALIB_ADC_DBG 0

// Maximum value to read from ADC (minimum is 0).
#define ADC_MAX 0xFFF

// Number of milliseconds to delay after turning on AREF before the reference
// voltage is stable. The larger the capacitor on pinAREF_OUT, the longer this
// should be.
#define AREF_STABLE_DELAY 5

// Number of milliseconds of delay after turning on PWM output before the
// voltage at the PWM capacitor is stable, in the worst case. This value can be
// determined with a scope and a section of code below, see comments below.
#define PWM_STABLE_DELAY 5

// Percentage of ADC range to use, e.g. if this is 10, measure the ADC value at
// steps between 10% and 90% of reference voltage and compute gain and offset
// corrections.
#define PERCENT_AT_ENDS 10

// Set this to 0 to disable all oscilloscope-related code. Otherwise, set it to
// an integer between 1 and 6 to enable a section of the calibration code that
// does a specific type of test that you can monitor on your scope:
//    1 - test pinPWM as a regular digital output pin. Watch it on a scope.
//    2 - test ADC values at 20% of max. Use a scope to watch the PWM waveform.
//          The PIN_DEBUG output is HIGH during the test.
//    3 - test PWM output rise time. Watch pinPWM and pinADC with scope,
//          triggering on PIN_DEBUG going high. It's a bit less than 3 ms.
//    4 - test PWM output rise time from 0% to PERCENT_AT_ENDS of max. Watch
//          pinPWM with scope, triggering on PIN_DEBUG going high. It's a bit
//          less than 5 ms from 0 to max voltage.
//    5 - test PWM output rise time from 100-PERCENT_AT_ENDS to 100% of max.
//          Watch pinPWM with scope, triggering on PIN_DEBUG going high. It's a
//          bit less than 2 ms.
//    6 - test ADC values at 0, quarter-max, half-max, 3/4-max, and max. The
//          PIN_DEBUG output is high during the test.
//  Work through these values one at a time and check signals to see that
//  everything looks good.
#define CALIB_ADC_SCOPE 0

// Arduino "pin" number of digital output pin to use for debug output (e.g. to
// trigger a scope). Used in conjunction with CALIB_ADC_SCOPE above. This is an
// index into g_APinDescription[] in variant.cpp. If CALIB_ADC_SCOPE is 0, set
// this to -1. On my system, when debugging I used a value of 5 = D5 = PA05.
#if CALIB_ADC_SCOPE > 0
#define PIN_DEBUG 5
#else
#define PIN_DEBUG -1
#endif

/**************************************************************************/
/*!
  @brief    Run a calibration algorithm on the ADC using the PWM output and TCC
            timer, compute ADC gain and offset constants, and load them into the
            ADC.
  @param    pinADC      Arduino "pin" number of an ADC-available analog input
                        pin to use to read the calibration voltage. This is
                        an index into g_APinDescription[] in variant.cpp, and
                        the table must include ADC_Channel# for this pin.
                        This pin has an 0.1 uF capacitor tied between it and
                        ground, and a 10K ohm resistor tied between it and
                        pinPWM. The default value is what I used on my system:
                          7 = D7 = PA06 = AIN6
  @param    pinPWM      Arduino "pin" number of digital output pin to use to
                        generate a PWM waveform using a SAMDG21 TCC timer. This
                        is an index into g_APinDescription[] in variant.cpp, and
                        the table must include TCC#_CH# and PIN_ATTR_PWM. This
                        pin has a 10K ohm resistor tied between it and pinADC.
                        Note: A TCC timer is used rather than a TC timer in
                        order that the period of the waveform can be set. The
                        default value is what I used on my system:
                          4 = D4 = PA07 = TCC1 ch 1
  @param    pinAREF_OUT Arduino "pin" number of digital output pin to use to
                        generate the ADC external reference voltage. This is an
                        index into g_APinDescription[] in variant.cpp, and the
                        table must include PIN_ATTR_DIGITAL. This pin has a
                        100uF capacitor tied between it and ground, and the pin
                        is also tied to the external AREF input. If you don'
                        want to use this pin but instead want to tie the AREF
                        input directly to 3.3V, set this to -1. The default
                        value is what I used on my system:
                          6 = D6 = PA04
  @param  cfgADCmultSampAvg   Set this to 0 to disable ADC automatic multiple
                              sampling and averaging, or a number X between 1
                              and 10 to average 2^X samples, e.g. 6 means
                              average 2^6 = 64 samples (seems to give good
                              results). This is internal hardware-based
                              averaging. It produces more stable ADC values.
  @note     Method: 33 evenly-spaced ADC samples are taken with ADC input (from
            a PWM circuit with capacitor and resistor) between 10% and 90% of
            the reference voltage of 3.3VDC. Least squares regression is used to
            fit the best curve to the relationship between the expected and
            actual ADC values, and the results of that provide an offset and
            gain value are loaded into the ADC gain and offset correction
            registers.
*/
/**************************************************************************/
extern void calibSAMD_ADC_withPWM(int pinADC = 7, int pinPWM = 4,
  int pinAREF_OUT = 6, uint8_t cfgADCmultSampAvg = 6);

#endif // calibSAMD_ADC_withPWM_h
